kedro-org/kedro
# Enable passing argument to line magic `%reload_kedro`

## Description

Close #1748 

It's not trivial to enable positional & keywords arguments with line magics, this PR aims to support:
* Accepting keywords arguments with the line magic, so user don't have to import the function `reload_magic` and have a more consistent UX.
* Keeping the existing behavior - passing `path` is optional.

## How to test
1. Open GitPod 
2. Check if you are in a kedro project (It should by default)
3. Try the following with `kedro ipython` / `kedro jupyter notebook`
```python
%reload_kedro    # Default path=None
%reload_kedro .  # With positional argument only
%reload_kedro . --env=base # Mix use of argument type
%reload_kedro --env=base # Keyword argument only, path default=None
```

## Development notes


* To make `%reload_kedro` and `reload_kedro` possible, `@register_line_magic("reload_kedro")` is needed to register a different function as the line magic with name `reload_kedro`
* `path` is made positional and `env` and `extra_params` are keywords arguments, partly due to compatibility issues. I thought about making it accepts both positional or keyword arguments, but it seems that this is not aligned with how CLI works in general, so I did not do it in the end.
  * There are some hints that how you can do that in this link.
https://stackoverflow.com/questions/14585722/can-you-pass-keyword-arguments-instead-of-positional-arguments-to-argparse


Todo:
- [x] Update docstring - `reload_kedro` + Shift-Tab in Jupyter should show up some examples. 
![image](https://user-images.githubusercontent.com/18221871/185198282-74e21d3d-678e-40b7-b8a8-212040a8ce65.png)

- [ ] Check `reload_kedro` without parameter still works correctly 
- [ ] Check `reload_kedro` with `path` as positional argument still works

### Side-notes (Only read if you have extra time)
https://ipython.readthedocs.io/en/stable/api/generated/IPython.core.interactiveshell.html?highlight=register_magic_function#IPython.core.interactiveshell.InteractiveShell.register_magic_function suggest that the magic should have a signature `f(line)` but this is not the case with `reload_kedro`. See an example here:
```python
def csv(line, cell):
    sio = StringIO(cell)
    return pd.read_csv(sio)

def load_ipython_extension(ipython):
    """This function is called when the extension is
    loaded. It accepts an IPython InteractiveShell
    instance. We can register the magic with the
    `register_magic_function` method of the shell
    instance."""
    ipython.register_magic_function(csv, 'cell')
```
Currently, we have `def reload_kedro(
    path: str = None, env: str = None, extra_params: Dict[str, Any] = None
)`, it is working only because it takes the first argument as `line` 

This PR use `reload_kedro_magic(line)` which call the `reload_kedro` method, so we keep the backward compatible that people can still import the same function. but `%reload_kedro` will call the line magic instead.

#### Note 3 
The `IPython` use `argparse` for the magic, it's less consistent with most of the `kedro` codebase which use `click`, but I didn't want to go all the way to re-invent the logic. In theory, it is an exercise of passing a long string into arguments and making sure it has the correct type.

## What's more to do?
* `extra_params` now use a `json.loads` method to construct the dictionary, I also consider if we should make it the same style as `kedro run --params` which involves `_split_params` and `_split_string`, but it is probably a rare case, so I would rather wait for a user request to actually implement this.


## Checklist


<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1789"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

# tests
`kedro ipython`, the log should be the same with this PR.
```
[08/17/22 08:37:40] INFO     Updated path to Kedro project: /workspace/project                      ipython.py:53
[08/17/22 08:37:40] WARNING  /home/gitpod/.pyenv/versions/3.8.12/lib/python3.8/site-packages/requ warnings.py:109
                             ests/__init__.py:102: RequestsDependencyWarning: urllib3 (1.26.6) or                
                             chardet (5.0.0)/charset_normalizer (2.0.4) doesn't match a supported                
                             version!                                                                            
                               warnings.warn("urllib3 ({}) or chardet ({})/charset_normalizer                    
                             ({}) doesn't match a supported "                                                    
                                                                                                                 
[08/17/22 08:37:41] INFO     Kedro project project                                                  ipython.py:76
                    INFO     Defined global variable 'context', 'session', 'catalog' and            ipython.py:77
```
